src/main/java/DukeTask/Task.java

```diff
@@ -0,0 +1,133 @@
+package duketask;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+/**
+ * Represents an abstract task storing the task description,
+ * state (whether it is done or not) and the time of creation
+ * by the end-user.
+ */
+public abstract class Task {
+    /**
+     * Exception thrown when the user does not provide any
+     * task description on construction.
+     */
+    public static class EmptyDescriptionException extends Exception {
+        private static final long serialVersionUID = -365632185953600895L;
+
+        public EmptyDescriptionException() {
+            super("task description cannot be empty! no sense...");
+        }
+    }
+
+    /**
+     * Exception thrown when user attempts to mark the task as done
+     * when the task is already in a done state.
+     */
+    public static class MarkedAsDoneException extends Exception {
+        private static final long serialVersionUID = -7397628460804553793L;
+
+        public MarkedAsDoneException() {
+            super("trickery! task has already been marked done by dogeDuke...");
+        }
+    }
+
+    /**
+     * Enum switch state for whether the task is done or not.
+     */
+    protected enum TaskState {
+        done, undone;
+    }
+
+    protected String description;
+    protected TaskState state;
+    protected LocalDateTime createdDateTime;
+
+    /**
+     * Class main constructor.
+     * @param description       the task description; cannot be empty.
+     * @param isDone            whether the task is done on creation.
+     * @param createdDateTime   the datetime at which the task is created.
+     * @throws Task.EmptyDescriptionException
+     */
+    public Task (String description, boolean isDone, LocalDateTime createdDateTime)
+            throws Task.EmptyDescriptionException {
+
+        if (description.isEmpty()) {
+            throw new Task.EmptyDescriptionException();
+        }
+
+        this.state = (isDone ? TaskState.done : TaskState.undone);
+        this.description = description;
+        this.createdDateTime = createdDateTime;
+    }
+
+    /**
+     * Class default constructor.
+     * passes the undone state and current datetime for the task isDone and
+     * createdDateTime parameters.
+     * @param description       the task description; cannot be empty.
+     * @throws Task.EmptyDescriptionException
+     */
+    public Task (String description) throws Task.EmptyDescriptionException {
+        this(description, false, LocalDateTime.now());
+    }
+
+    // accessors
+    public boolean isDone () {
+        return (this.state == TaskState.done);
+    }
+
+    public boolean inDescription (String keyphrase) {
```

Comment:
Do you think a better alternative for this method name would be `isInDescription`?